### PR TITLE
Validate VRF for inner messages & optimization

### PR DESCRIPTION
**Pull request by gavraz**
* Make sure VRF is checked also for inner messages in SVP & certificate
* Optimized inner messages validation with valid messages cache

### CODE REVIEW

File: hare/algorithm.go [hunk 1]
```diff
@@ -312,18 +317,15 @@ func (proc *ConsensusProcess) onEarlyMessage(m *Msg) {
...
-	mType := messageType(m.InnerMsg.Type).String()
-	// validate InnerMsg for this or next round
+	// validate context
 	err := proc.validator.ContextuallyValidateMessage(m, proc.k)
 	if err != nil {
-		if err == errEarlyMsg { // early message, keep for later
+		mType := messageType(m.InnerMsg.Type).String()
+
+		// early message, keep for later
+		if err == errEarlyMsg {
```
Comment (zalmen) [inline hare/algorithm.go:330]: you are not syntactically validating an early message. This basically gives an (eligible) attacker a direct (write) access to the node's memory

File: hare/algorithm.go [hunk 2]
```diff
@@ -152,10 +152,14 @@ type ConsensusProcess struct {
 	cfg               config.Config
 	pending           map[string]*Msg // buffer for early messages that are pending process
 	notifySent        bool            // flag to set in case a notification had already been sent by this instance
+	mTracker          *msgsTracker    // tracks valid messages
 }
 
 // NewConsensusProcess creates a new consensus process instance.
-func NewConsensusProcess(cfg config.Config, instanceId instanceId, s *Set, oracle Rolacle, stateQuerier StateQuerier, layersPerEpoch uint16, signing Signer, nid types.NodeId, p2p NetworkService, terminationReport chan TerminationOutput, logger log.Log) *ConsensusProcess {
+func NewConsensusProcess(cfg config.Config, instanceId instanceId, s *Set, oracle Rolacle, stateQuerier StateQuerier,
```
Comment (y0sher) [inline hare/algorithm.go:159]: note to the future. it seems that you have a lot of constructor configurations. I'd pass a single `ConsensusProcesesConfig` struct that holds all the information needed to start a process.

Comment (gavraz) [reply hare/algorithm.go:159]: Ok, but do you mean a config that is a collection of all primitives? For example, a set is not a primitive... or do you mean just a single config that includes all?

File: hare/messagevalidation.go [hunk 3]
```diff
@@ -219,91 +234,111 @@ func (validator *syntaxContextValidator) SyntacticallyValidateMessage(m *Msg) bo
...
+			senders[pub.String()] = struct{}{} // mark sender as exist
+
+			// passed validation, continue to next message
+			continue
 		}
 
-		if !validator.SyntacticallyValidateMessage(iMsg) {
-			validator.Warning("Aggregated validation failed: identified an invalid inner message %v", iMsg)
-			return false
+		// extract public key
```
Comment (barakshani) [inline hare/messagevalidation.go:295]: this is more than just extracting, you check activation as well

File: hare/messagevalidation.go [hunk 4]
```diff
@@ -219,91 +234,111 @@ func (validator *syntaxContextValidator) SyntacticallyValidateMessage(m *Msg) bo
...
 
-		// validate unique sender
 		pub := iMsg.PubKey
+		// validate unique sender
 		if _, exist := senders[pub.String()]; exist { // pub already exist
-			validator.Warning("Aggregated validation failed: detected same pubKey for different Messages")
-			return false
+			return errDupSender
 		}
 		senders[pub.String()] = struct{}{} // mark sender as exist
```
Comment (barakshani) [inline hare/messagevalidation.go:306]: update v.validMsgsTracker as well

**Review by barakshani — APPROVED** [review]

**Review by barakshani — APPROVED** [review]

**Review by zalmen — APPROVED** [review]

**Pull request merged**